tests/deploy/ui/test_ui.py

```diff
@@ -0,0 +1,183 @@
+import os
+import time
+from multiprocessing import Process
+
+import httpx
+import panel as pn
+import pytest
+import uvicorn
+from playwright.sync_api import expect, sync_playwright
+
+from ragna.assistants import RagnaDemoAssistant
+from ragna.core._utils import default_user
+from ragna.deploy import ApiConfig, Config, UiConfig
+from ragna.deploy._api import app as api_app
+from ragna.deploy._ui import app as ui_app
+
+TEST_API_PORT = "38769"
+TEST_UI_PORT = "38770"
+
+
+class TestAssistant(RagnaDemoAssistant):
+    def answer(self, prompt, sources, *, multiple_answer_chunks: bool):
+        content = next(super().answer(prompt, sources))
+
+        if multiple_answer_chunks:
+            for chunk in content.split(" "):
+                yield f"{chunk} "
+        else:
+            yield content
+
+
+@pytest.fixture(scope="session")
+def headed_mode(pytestconfig):
+    return pytestconfig.getoption("headed") or False
+
+
+@pytest.fixture
+def config(tmp_local_root):
+    return Config(
+        local_root=tmp_local_root,
+        assistants=[TestAssistant],
+        ui=UiConfig(port=TEST_UI_PORT),
+        api=ApiConfig(port=TEST_API_PORT),
+    )
+
+
+class ApiServer:
+    def __init__(self, config):
+        self.config = config
+
+    def start_server(self):
+        uvicorn.run(
+            api_app(
+                config=self.config,
+                ignore_unavailable_components=True,
+            ),
+            host=self.config.api.hostname,
+            port=self.config.api.port,
+        )
+
+    def server_up(self):
+        try:
+            return httpx.get(self.config.api.url).is_success
+        except httpx.ConnectError:
+            return False
+
+    def start(self):
+        self.proc = Process(target=self.start_server, args=(), daemon=True)
+        self.proc.start()
+
+        timeout = 5
+        while timeout > 0 and not self.server_up():
+            print(f"Waiting for API server to come up on {self.config.api.url}")
+            time.sleep(1)
+            timeout -= 1
+
+    def stop(self):
+        self.proc.kill()
+
+
+@pytest.fixture
+def api_server(config):
+    server = ApiServer(config)
+    try:
+        server.start()
+        yield server
+    finally:
+        server.stop()
+
+
+def auth_header(base_url):
+    username = default_user()
+    token = (
+        httpx.post(
+            base_url + "/token",
+            data={
+                "username": username,
+                "password": os.environ.get(
+                    "RAGNA_DEMO_AUTHENTICATION_PASSWORD", username
+                ),
+            },
+        )
+        .raise_for_status()
+        .json()
+    )
+
+    return f"Bearer {token}"
+
+
+@pytest.fixture(scope="function")
+def page(config, api_server, headed_mode):
+    server = ui_app(config=config, open_browser=False)
```

Comment:
This was one of my questions I forgot to include.

When I had the panel `.serve()` call outside of the `sync_playwright()` scope it would just hang forever. I couldn't figure out why so it ended up in this block.

Initially both the UI server and the API server were just plain functions with no class wrapper, but both MacOS and Windows were struggling to pickle the closures during the pytest setup so I moved the API server to its own class. I will first try to use the CLI func for the UI and fallback to making individual classes for both the UI and API for consistency.